components/service/glean/src/main/java/mozilla/components/service/glean/timing/TimingManager.kt

```diff
@@ -0,0 +1,107 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.timing
+
+import android.os.SystemClock
+import mozilla.components.service.glean.error.ErrorRecording
+import mozilla.components.service.glean.error.ErrorRecording.recordError
+import mozilla.components.service.glean.private.CommonMetricData
+import mozilla.components.support.base.log.logger.Logger
+import java.util.WeakHashMap
+
+/**
+ * A class to record timing intervals associated with a given arbitrary object.
+ *
+ * A timings are recorded and returned in nanoseconds.
+ */
+internal object TimingManager {
+    private val logger = Logger("glean/TimingManager")
+
+    /**
+     * A map that stores the start times of running timers.
+     *
+     * A [WeakHashMap] is used so that we don't unintentionally leak memory
+     * by keeping references to otherwise destroyed objects around.
+     */
+    val uncommittedStartTimes = mutableMapOf<String, WeakHashMap<Any, Long>>()
+
+    /**
+     * Helper function used for getting the elapsed time, since the process
+     * started, using a monotonic clock.
+     * We need to have this as an helper so that we can override it in tests.
+     *
+     * @return the time, in nanoseconds, since the process started.
+     */
+    internal var getElapsedNanos = { SystemClock.elapsedRealtimeNanos() }
+
+    /**
+     * Start tracking time associated with the provided object. This records an
+     * error if it’s already tracking time (i.e. start was already called with
+     * no corresponding [stop]): in that case, the original start time will be
+     * preserved.
+     *
+     * @param metricData The metric managing the timing. Used for error reporting.
+     * @param anyObject The object to associate with this timing.
+     */
+    public fun start(metricData: CommonMetricData, anyObject: Any) {
+        val startTime = getElapsedNanos()
+        val metricName = metricData.identifier
+
+        uncommittedStartTimes[metricName]?.let { metricTimings ->
```

Comment:
Should this access be `synchronized` as well?

Reply:
I was thinking since it's read-only maybe not, but I think you're right...

Also, this raises a concern about blocking on these.  I think I need to rework the async here so these updates happen off the main thread...